### DISCUSSION TITLE
With the script, I always have 600 watts going into the power grid

### DISCUSSION THREAD

**Discussion by MrScotch72** [svenerbe/ecoflow_dynamic_power_adjustment#28] [Q&A] [answered]
I have installed the script.
Entered the values ​​for key and secret.
When I start the automation, the Powerstream draws 600 watts from the Delta 2 battery
 (it is currently 11:30 p.m. and no sun ;-) ). 
 However, my current consumption is around 230 watts. (It always fluctuates a bit)
This is the code from my automation.

`service: pyscript.set_ef_powerstream_custom_load_power
data:
  SerialNumber: HW51...........30
  TotalPower: >-
    {{ (states('sensor.tasstrom_zahler_16_7_0') | float) +
    (states('sensor.ef_ps_800_inverter_output_watts') | float) -2 | round }}
  Automation: true`
  
 When I test the code in the template editor, the current consumption is displayed. 
<img width="1330" alt="Bildschirmfoto 2024-07-08 um 23 33 33" src="https://github.com/svenerbe/ecoflow_dynamic_power_adjustment/assets/83172846/81f90cdb-6ec6-418e-9316-06b98303937c">

**Comment by Just-Zuul** [comment]
What is your preferred language?

No wonder you feed in full, you do not take into account the current inverter value correctly

TotalPower: >- {{ (states('sensor.tasstrom_zahler_16_7_0') | int) - (states('sensor.ef_ps_800_inverter_other_loads') | int) -20}} Automation: true

When (states('sensor.tasstrom_zahler_16_7_0') | int) is your Power Supply and  (states('sensor.ef_ps_800_inverter_other_loads') | int) is your Inverter (not the full Inverter Output - only the other_load (without SmartPlug Value) ) 

Tip: subtract 20 or better 50 watts from the feed-in power, otherwise you feed in more than you save because the meters jump back and forth depending on consumption and otherwise you'll always end up below zero

This works for me with 3 inverters at the same time. I have 3 DeltaPro and 3 Powerstrems with 7 panels registered as one system and although each inverter does its own thing, they coordinate perfectly with each other (3 WLAN / 1 dev account)

By the way - thanks to Sven for the gigantic script - it's unbelievable what you can do with it

**Reply by MrScotch72** [reply]
Thank you for your answer.
My preferred language is German, if I understood your question correctly.
I have adapted the code according to your instructions.
(changed float to int) removed the round (which is then superfluous) and adjusted the -2 to -20.
You write "sensor.ef_ps_800_inverter_other_loads" in your answer
For me it says "sensor.ef_ps_800_inverter_output_watts"
is ef_ps_800_inverter_other_loads a helper for you? When I look at the entities from the PowerStream (I have 57 listed) in Home Assistant, I cannot find ....inverter_other_loads there.
The sensor "sensor.tasstrom_zahler_16_7_0" shows the current power consumption of my meter. This value can be positive (purchase from energy supplier) or negative (feeding in too much solar power).
If I activate your changes in the automation, it will be set back to 600 watts.
---------
Danke für Deine Antwort.
Meine bevorzugte Sprache ist Deutsch, Du darfst mir auch gerne auf Deutsch antworten ;-) 
Den Code habe ich entsprechend Deiner Hinweise angepasst.
(float in int geändert) das round entfernt (ist ja dann überflüssig) und die -2 auf -20 angepasst.
Du schreibst in deiner Antwort "sensor.ef_ps_800_inverter_other_loads"
Bei mir heist es ja "sensor.ef_ps_800_inverter_output_watts"
ist ef_ps_800_inverter_other_loads ein Helfer bei Dir? Wenn ich mir die Entitäten von der PowerStream anschaue (ich habe 57 gelistet) bei Home Assistant, dann finde ich dort ....inverter_other_loads nicht.
Der Sensor "sensor.tasstrom_zahler_16_7_0" zeigt den aktuellen Stromverbrauch von meinem Zähler an. 
Dieser Wert kann positiv (Bezug vem Energieversorger) oder negativ (Einspeisung von zu viel Solarstrom) sein. 
Wenn ich Deine Änderungen aktiviere in der Automatisation, wird wieder auf 600 Watt gestellt.

**Comment by Just-Zuul** [comment]
OK - Deutsch - das ist einfacher für mich ;-)

Nein es ist kein Helfer es ist ein Sensor des Inverters. Sorry - ich habe mit dem Script Tonnen an verschiedenen Automatisationen und Konstellationen - und dabei so viel angepasst - wird langsam verwirrdend.

Lass mich kurz mal zusammenfassen:
Du hast einen PowerStream, einen Akku dran, hast HomeAssistant mit dem tolwi/hassio-ecoflow-cloud - repository eingerichtet. 
In dieser Integration hast Du den Inverter eingerichtet (soweit korrekt?). Hast Du SmartPlugs? Wenn Ja / Nein - egal - aber hast Du den Sensor  _smart_plug_loads ?

Der Powerstream hat 3 Angaben (Entitäten / Sensoren) für Output - wenn die Entitäten aktiviert sind (oder wurden) dann werden sie in HA  als Sensoren gesehen.

Nehmen wir an, Dein Inverter hat in der App bei der Einrichtung den Namen PowerStream bekommen (oder belassen). Der Name steht fest, BEVOR Du ihn in HA integrieren möchtest.

Dann sieht das so aus

- sensor.PowerStream_inverter_output_watts

- sensor.PowerStream_smart_plug_loads

- sensor.PowerStream_other_loads


1 ) - inverter_output_watts = der Wert des (angestrebten) gesamten Ausgang des PowerStreams - der ergibt sich aus der Summe von:

2 ) - smart_plug_loads = das, was Deine SmartPlugs anfordern

3 ) - other_loads = ist das was in Deiner App als Grundlast angegeben ist

![grafik](https://github.com/svenerbe/ecoflow_dynamic_power_adjustment/assets/153944468/e51f5039-fbaa-41e2-8ff4-e6ebd54c71f3)


Das Steuerungsscript ziehlt auf die Einstellung / Anpassung der Grundlast ab. Dadurch, dass Du den Wert der Grundlast, die ja in der Regel statisch ist, über das Script ständig anpasst, wird der Lastausgleich erreicht.

Das Aktionsscript in Deiner Automation macht nichts anderes als den Wert von Deinem Stromzähler MINUS dem Wert der eingestellten Grundlast (- ausgleich) zu nehmen und es über das pyscript in die Cloud zu schicken.

Da wird, wenn du gerade 300 Watt Last auf dem Zähler hast, nicht 300 übermittelt sondern nur die in diesem Moment bestehende Differenz zu Deinem Inverter (der eingestellten Grundlast - also dem other_load Wert) - also z.B. Werte wie -3 8 4 17 -20 -4 10 19 200 -60 usw.... 

Ich müsste mich schon wahnsinnig irren oder vergessen haben, dass / ob ich die Sensoren selber umbenannt habe aber Fakt ist, Du musst diese 3 Sensoren haben (aktiv oder inaktiv).

Wenn Du unter Entwicklerwerkzeuge -> Zustände in "Entitäten filtern" mal "other" eingibst - kommt da was bei Dir?

**Reply by MrScotch72** [reply]
Moin,
ich habe gerade geschaut. 
Alle drei Sensoren habe ich gefunden.

sensor.ef_ps_800_inverter_output_watts
sensor.ef_ps_800_other_loads
sensor.ef_ps_800_smart_plug_loads

Ich habe es jetzt auf 
`
  TotalPower: >-
    {{ (states('sensor.tasstrom_zahler_16_7_0') | int) -
    (states('sensor.sensor.ef_ps_800_other_loads') | int) -20 }}`
    
geändert. Da mein Geschirrspüler gerade läuft (SmartPlug) sehe ich noch kein unterschied. 
Außerdem ist jetzt die Sonne auch schon weg. 
Ich werde es morgen mal im Auge behalten. 

Wenn ich das im Template Editor so eingebe, bekomme ich eine Fehlermeldung. 

<img width="1266" alt="image" src="https://github.com/svenerbe/ecoflow_dynamic_power_adjustment/assets/83172846/a5202eff-c016-47ab-97d0-6882bd28acc7">

**Reply by Just-Zuul** [reply]
Du hast auch "sensor" 2 x drin....
(states('sensor.sensor.ef_ps_800_other_loads')

Gib mal in Deinem Template Editor das hier rein:

Stromzähler      : {{ (states('sensor.tasstrom_zahler_16_7_0') | int) }}
Grundlast  -0    : {{ (states('sensor..ef_ps_800_other_loads') | int) }}
Grundlast  -20  : {{ (states('sensor..ef_ps_800_other_loads') | int) -20 }}
------------------------------------------------------------------------
TotalPower (Differenzwert) : {{ (states('sensor.tasstrom_zahler_16_7_0') | int) - (states('.ef_ps_800_other_loads') | int) -20 }}

Lass Dich von dem Begriff TotalPower nicht verwirren - es ist ein Differenzwert !

**Reply by Just-Zuul** [reply]
ahh - Schreibfehler....

Stromzähler : {{ (states('sensor.tasstrom_zahler_16_7_0') | int) }}
Grundlast -0 : {{ (states('sensor..ef_ps_800_other_loads') | int) }}
Grundlast -20 : {{ (states('sensor..ef_ps_800_other_loads') | int) -20 }}
TotalPower (Differenzwert) : {{ (states('sensor.tasstrom_zahler_16_7_0') | int) - (states('sensor.ef_ps_800_other_loads') | int) -20 }}
--------------------------------------------------------------
manchmal....

**Reply by MrScotch72** [reply]
Hab ich gemacht.
sieht dann so aus.
<img width="1116" alt="Bildschirmfoto 2024-07-11 um 16 19 53" src="https://github.com/svenerbe/ecoflow_dynamic_power_adjustment/assets/83172846/31bacea0-1bf1-44a2-8f14-8a3eee6a705f">
Du hattest hinter dem Sensor und dem ef ein Punkt zu viel. 

welche Updaterate hast Du in Deiner Automatisation eingestellt?

**Reply by Just-Zuul** [reply, marked as answer]
na sieht doch super aus - auch ich darf doch mal blind sein - immerhin kein ganzer Sensor sondern nur sein Punkt  ;-)

4 Sekunden hab ich für die Automatisierung - mein Tasmota aktualisiert jede Sekunde, kannst du im Sensorscript einstellen

Aber ich habe hier auch scripttechnisch etwas die Sau rausgelassen - die Automatisation hat bei mir 3 Optionen in der Aktion je nachdem wie viel Sonne auf die Paneele kommt, etliche kombinierte Bedingungen, Schaltungen und noch ein paar "Sicherungen" damit ich nicht bei Schwankungen zu viel ins Netz baller.

Du kannst mit dem Script auch super eine feste Einspeisung machen und dann eine Umschaltung ob fest oder dynamisch die Grundlast angepasst werden soll - so für Tage mit wenig Solar, Herbst / Winter und ein bischen hamstern für den Notfall ;-)

input_number. FELD und dann statt Schieber ein Eingabefeld, Grenze 0-800 - den input_number statt dem Stromzähler und bumm haste ein Eingabefeld für den festen Grundbedarf.
Die Automatisattionsaktion ist dann (Schema):
`  TotalPower: >-
    {{ (states('input_number.wr_grundeinspeisung') | int ) - (states('sensor.wr_other_loads') | int ) }}`

Ich brauche so gut wie nie die App - geht alles wichtige über HA bis hin zu Ladestand - Ober/ Untergrenzen laden / entladen und die Speicher machen damit auch keine Probleme wegen ausbleibender Reaktion auf Morgensonne - meine 3 laufen seit 2 Monaten durch.
Man kann da sooo viel Zeit drin versenken.

Ach ja - das mit dem Abzugswert (die minus 20) - auch das kannst Du mit einem Nummerneingabefeld (input_number.abzugblabla) lösen und nach Bedarf anpassen. Du wirst merken dass, wenn Du Geräte hast die Schwanken (Pulsierende Verbraucher zB. Waschmaschine, Heizdecke wenn Frau kalte Füße hat, Ladegeräte...), dann springt dir der Wert ständig von über Null nach unter Null und am Tagesende sind 200 - 400 Wh im Netz verschwunden - mit einem Eingabefeld kannst Du feinjustieren was Du abziehen möchstest und dann gehen halt vom Netz so 50/60 Watt ins Haus und der Rest macht der Inverter mit der Batterie.

Wenn Du die dynamische Einspeisung etwas justieren willst.... ab 700 Watt Netzbezug nimmt er den vollen Wert, darunter den Wert minus dem Abzug:
`  TotalPower: |
    {% if states('sensor.tasmota_power_curr') | float >= 700 %}
      {{ states('sensor.tasmota_power_curr') | int }}
    {% else %}
      {{ (states('sensor.tasmota_power_curr') | int) -
         (states('input_number.abzugsjustierung') | int) }}
    {% endif %}`

Dann mal viel Erfolg und viel Spaß beim spielen - hast ja jetzt alles was du brauchst....

**Reply by MrScotch72** [reply]
Auf jeden fall erst mal vielen Dank:
Mittlerweile scheint es zu laufen.  
![IMG_7756](https://github.com/user-attachments/assets/d326168d-989e-45d4-9d3f-3a3c779f99cd)
Das mit der Anpassungen hört sich interessant an. Muss ich mir noch mal näher anschauen.

**Reply by Just-Zuul** [reply]
Cool - schön, dass Du es geschafft hast...

Wo ich es gerade sehe - nur mal so am Rande - den Speicher auf unter 10 zu fahren ist nicht soo toll für die Lebenserwartung der Zellen.

Den Fehler habe ich mit meiner ersten DeltaPro auch gemacht. Die hat jetzt 136 Cycles in 1,5 Jahren und ist im Mai bei 108 Cycles im State of health schon von 100 auf 98% gefallen - gar nicht gut ...

Versuche die Batterie im Bereich 20-80% zu fahren. 
Das wird von EcoFlow auch empfohlen, - klar - die wissen warum. Die Teile sind jetzt nicht das mega-high-end-Produkt. Ist eh alles subventionierter billig produzierter China Kram, überteuert auf den EU Markt geworfen.

15-85 als absolute Grenzen - und schauen, dass sie nicht zu lange immer an den Grenzen liegt. Wenn Du mehr reinbekommst und nicht weißt wohin mit dem Strom, lieber mal schauen ob Du bei einer Aktionswoche einen Zusatzakku zum guten Kurs abgreifen kannst.
(Klugscheißermodus aus)

Der Kram macht eh süchtig - mit einem Inverter und 2 Panelen angefangen und nun - boah eh...

Na dann - weiterhin viel Erfolg und Spaß mit der Sache.

**Reply by MrScotch72** [reply]
Ja das Stimmt mit dem Süchtig machen.
Zusatzakku ist schwierig be der Delta 2. Überlege schon mir eine Delta 2 Max zu holen.  
Hab das auch mal mit dem Umschalten zwischen dynamische und manuelle Einstellung probiert. 
Technisch klappt es, optisch brauch ich noch mal etwas Inspiration. 
Magst Du mal dein Dashboard zeigen?

**Reply by Just-Zuul** [reply]
Ich kann gerne versuchen mein derzeitiges Dashboard hier abzubilden - schätze aber, dass ist etwas verwirrend...
Ist noch nicht final - nur bisher um erstmal alle Steuerungen und Abgleiche zu erstellen und die Auswirkungen im Blick zu haben.

![grafik](https://github.com/user-attachments/assets/ae95bc5b-c9d6-4658-ac6d-b74490eabd1c)
![grafik](https://github.com/user-attachments/assets/38a3a017-ce82-46ab-a968-ed86b95974db)
![grafik](https://github.com/user-attachments/assets/600e1f6a-b1d7-4b63-9848-ae38ccb697cc)

Mal ein Teil des momentanen Dashboards - 

Ich habe so gut es geht alle Einstellmöglichkeiten als Input_number Felder definiert, da ich mit Automatisationen die Werte darin je nach Gegebenheit anpassen kann.

Wenn z.B DP 3 auf einen Wert höher 4 über dem Akkustand ankommt, wird der Wert aus der dynamischen Einspeisung automatisch der DP2 / WR 2 zugeordnet und der Abzugswert (den ja im Moment die 3 liefert) entfernt.

Das auch deswegen, weil die DP3 und der WR3 hauptsächlich die Last meiner Klimaanlage einspeist - meist reicht der Inhalt für 24 Stunden bzw. über Nacht (in heißen Nächten)...

DP 1 mit WR 1 decken den Rest ab, den die 2 & 3 nicht bringen.

Da ich an jedem Außenleiter (L1, L2, L3) jeweils direkt einen Wechselrichter dran habe (WR1,2,3) habe ich mit Lastausgleich nicht so gewaltige Verschiebungsprobleme.

Dadurch, das die SmartPlugs dauernd aus der App rausfliegen (aus dem WLAN - die Dinger sind echter Murks) habe ich einen Großteil der WLAN-Steckdosen die hier und da angebracht sind in einem Sensor kombiniert (siehst du in dem Feld zusätzliche Einspeisung zur statischen).

Der Wert geht über eine Automatisation alle 4 Sekunden in das Feld "Dynamische Grundeinspeisung" (ich kann über die Schalter den Wert entweder dem WR2 oder dem WR 3 zuordnen) das heißt, ich kann jede x-beliebige Steckdose als SmartPlug nehmen .

Die größten Probleme macht eh die EcoFlow Cloud die bei Amazon auf dem AWS liegt. keine Ahnung wo / bei wem die Probleme sind, aber wenn das ganze ohne externe Cloud laufen könnte - das wäre mal geil.

Leider sind die Entwickler von HA ja nicht gerade auf custom Integrations begeisterungswürdig zu stimmen. Auch so ein Punkt der echt schade ist - es ist so derart unglaublich was alles möglich wäre / ist und noch sein könnte.

Ich wusste nicht was Du für einen Akku hast - ich hatte anfangs auch lange rumüberlegt aber mich dann doch für die DeltaPro (DP) entschieden - auch wegen der Kapazität. die 2. und 3. habe ich mir in Aktionen geschossen und daher einen recht guten Kurs gehabt. Plus Punkte oder Aktionsrabbatcodes und da sind dann schon mal 900 Euronen weniger geflossen pro Set mit dem WR.

Die Delta 2 hat glaube ich auch Lion Akkus... war nicht soo überzeugend für mich und ich wollte LFP's.

Wenn es mir nicht reicht kann ich immer noch je einen Akku zusätzlich an jede DP hängen...

Ja - ja - ich weiß - Amortisation - ich höre es schon - für mich amortisiert sie sich wenn sie angeschlossen ist und ich mit 1 - 3 kWh über den Tag komme, trotz Klimaanlage, Waschmaschine, Spülmaschine und HomeOffice (statt sonst 15 - 20 kWh).
Und wer weiß schon was noch kommt....

Wie gesagt - auch hier ist nochts final fertig aber es läuft im Moment ganz gut. Nur die Zeit die man darin versenkt ist manchmal erschreckend. Mache selbst erst seit Februar mit HA rum und das kann schön anspruchsvoll sein. Aber es macht auch spaß.....
Bis dann....

**Reply by MrScotch72** [reply]
Danke fürs teilen. 
Über Amortisation rede ich auch nicht. 
Ich Sehe es als Hobby an. Hobby kostet... 
Die Delta 2 hat keine LiPo sonder schon LiFePo.
Darf ich blöd fragen wie du es angemeldet hast? 
Wenn ich es richtig sehe hast Du 6 Platen an drei WR laufen.
Bin auch am überlegen noch 2 Platten drauf zu packen. 
Hab ja noch eine DEYE 600 im Keller liegen.

**Reply by Just-Zuul** [reply]
Es sind 7 Platten : 
2 an WR1, 3 an DP2 und 2 an DP3 oder DP1 je nach Jahreszeit / Wetter. 
Ich suche da noch eine akzeptable Lösung über HA zu steuern, ob die letzten 2 Platten in die 1 oder 3 Speisen (DC Schütz-Schaltung oder ähnliches). Kannst ja an der DP direkt ein paar Platten anschließen und 1200 Watt reichen da völlig am XT60 Anschluss.
Am WR 2 und 3 sind noch keine Platten angeschlossen - aber schonmal vorbereitet.

Angemeldet hab ich es mit dem freundlichen Elektriker aus dem Dorf. Ist ein bischen Formulare ausfüllen - dauert 30 Minuten wenn man gemütlich macht.

Was einem ja keiner sagt : 
Speicher sind nicht in der vereinfachten Anmeldung beim Netzbetreiber enthalten - einen Speicher muss man immer mit einem Elektriker anmelden. Du kannst zwar den Speicher bei der Anmeldung beim Marktstammdatenregister mit angeben, aber dann kommt der NB und fordert die Anmeldung über den Elektriker. 
Zumindest hat mein Netzbetreiber darauf bestanden. Ich hatte das 1. BKW mit Speicher beim Marktstammdatenregister eingetragen und schon kam der Netzbetreiber und meinte ich soll über einen Elektriker den Speicher anmelden.

Also dachte ich mir - ok - dann hol ich mir noch 2 Anlagen und melde alles als eine an.

Mein Netzbetreiber scheint da eh abgeneigt zu sein, wenn die Kunden ihren Strom teilweise selber machen.

**Comment by Neewieedo** [comment]
Da hier in deutsch geschrieben wird, mache ich einfach mal so weiter... ;-)

Ich hab zu dem Thema folgende Fragen:

1. In welcher Frequenz / Zeit sendet Ihr neue Werte an den Wechselrichter?

2. Wie bildet Ihr aktuell den Wert den Ihr an der Wechselrichter sendet? Also erfolgt die Berechnung außerhalb des pyscripts und wenn ja wie habt Ihr das mit dem aktuellen pyscript Code gelöst, da dieser die Berechnung ja auch durchführt:
`CalPermanentWatts = cur_permanentWatts + TotalPower - TotalPowerOffSet`

3. Gibt es eine Möglichkeit die keys und secret aus dem pyscript in die secrets.yaml zu "verlinken"?

4. Wie stellt Ihr sicher, das Ihr aktuelle / valide Strombezugswerte von Tibber habt?
Bei mir fieren die Werte immer wieder ein und das führt dann zu einer statischen / falschen Einspeisung über den Powerstream.
Ich habe es versucht über folgende Automation zu lösen, aber irgendwie wird der Schalter nicht ausgelöst...
```
alias: Tibber Status ermitteln
description: ''
trigger:
  - platform: state
    entity_id:
      - sensor.power_tibber
    for:
      hours: 0
      minutes: 1
      seconds: 0
    id: '1'
  - platform: state
    entity_id:
      - sensor.power_tibber
    to: unavailable
    id: '1'
  - platform: state
    entity_id:
      - sensor.power_tibber
condition: []
action:
  - if:
      - condition: trigger
        id:
          - '1'
    then:
      - service: switch.turn_off
        metadata: {}
        data: {}
        target:
          entity_id: input_boolean.tibber_status
    else:
      - service: switch.turn_on
        metadata: {}
        target:
          entity_id: input_boolean.tibber_status
mode: single
```
Ich hoffe, ihr habt den ein oder anderen Tipp für mich :-)

**Comment by MrScotch72** [comment]
Moin,
ich aktualisiere z.Z. alle 15 Sec. in der Automatisation. 
Das klappt so la la, ich denke man könnte noch etwas runter gehen mit der Zeit. 

Zum Auslesen habe ich einen IR Lesekopf mit Tasmota 
Darüber gebe ich die Daten per mqtt an HA.
Das ist die Automatisation die alle 15 sec. aufgerufen wird. 
`service: pyscript.set_ef_powerstream_custom_load_power
data:
  SerialNumber: HWxxxxxxxxxF83xx
  TotalPower: >-
    {{ (states('sensor.tasstrom_zahler_16_7_0') | int) -
    (states('sensor.ef_ps_800_other_loads') | int) -20 }}
  Automation: true`
  
  Der sensor.tasstrom_zahler_16_7_0 is der Aktuelle Verbrauch vom Zähler
    
![image](https://github.com/user-attachments/assets/ee11faf9-297b-4814-bea7-84395b9579f2)

Zu den Keys kann ich leider nichts sagen, genauso wenige zum Tibber, da ich ja mit Tasmota arbeite. 
Hoffe es hilft Dir weiter.

LG

**Comment by Neewieedo** [comment]
Hi,

vielen Dank für deine Rückmeldung.

Allerdings ist mit nicht ganz klar, wie das funktionieren soll... Wenn Dein Tasmotazähler einen zuverlässigen Wert liefert, solltest Du mal folgendes probieren:

`service: pyscript.set_ef_powerstream_custom_load_power data: SerialNumber: HWxxxxxxxxxF83xx TotalPower: >- {{ (states('sensor.tasstrom_zahler_16_7_0') | int) - 20 }} Automation: true`

Dann hättest Du immer 20 Watt Strombezug als "Sicherheit" das nichts in Netz zurück geht.

Die aktuelle Einspeisung durch den Powerstream verrechnet ja wie oben erwähnt bereits das pyskript:
`CalPermanentWatts = cur_permanentWatts + TotalPower - TotalPowerOffSet`

Gruß

**Reply by Just-Zuul** [reply]
@Neewieedo 
Deine Angabe ist grundsätzlich richtig und funktioniert auch. 
Ich hatte es früher auch so. Das _kann_ aber Probleme machen, in der Reaktionszeit vom Powerstream.

Wenn ich es richtig verstanden habe, bezieht das Script den aktuellen Wert erstmal aus der Cloud und berechnet dann den Änderungswert den es dann zurückschickt (laienhaft ausgedrückt).

Wenn Du es aber so machst:
 "Stromzähler - other_loads"
 dann berechnest Du schon den Unterschied direkt VOR der Übergabe anhand Deiner Werte ohne Cloudbezug.

In der Regel sind die Werte in beiden Gestaltungen am Ende immer gleich. 
Außer die Cloud antwortet nicht zuverlässig oder schnell genug.

**Nur falls es von Interesse ist:**
Da der Inverter nicht immer gleich hinterherkommt, besonders wenn er bei Null startet, habe ich mein Script in der Automatisation für den TotalPower so gestaltet:

service: pyscript.set_ef_powerstream_custom_load_power
data:
  SerialNumber: HWxxxxxxxxxxxxxxxxx
  TotalPower: >
    {% set tasmota_value = states('sensor.tasmota_lesekopf_power_curr') | int %}  
    {% set abzugswert = states('input_number.globaler_abzug') | int %}
    {% set schwelle = states('input_number.schwellwert') | int %}
    {% set inverter_ist = states('sensor.shelly_real_einspeisung_watt') | int %}
    {% set inverter_soll = states('sensor.inverter_output_watts') | int %}
    {% set ausgleichswert = states('sensor.inverter_other_loads') | int %}
    {% set inverter_diff = inverter_soll - inverter_ist %}
    {% set acceptable_range =
      7 if inverter_soll < 50 else
      10 if inverter_soll < 150 else
      15 if inverter_soll < 300 else
      20 if inverter_soll < 400 else
      25 %}
    {% if tasmota_value < schwelle %}
      {% set diff = schwelle - tasmota_value %}
      {% set neuer_abzugswert = abzugswert + diff %}
      {% if neuer_abzugswert > 800 %}
        {% set neuer_abzugswert = 800 %}
      {% elif neuer_abzugswert < 0 %}
        {% set neuer_abzugswert = 0 %}
      {% endif %}
      {{ tasmota_value - neuer_abzugswert }}
    {% else %}
      {{ tasmota_value - abzugswert }}
    {% endif %}
  Automation: true
enabled: true

    
hier wird berücksichtigt, welchen Wert der Inverter "wirklich" bereits abgibt und die Einspeisung wird darauf ausgerichtet, dass der Wert nicht übers Ziel hinausschießt, bzw. zu sehr hin und her schwankt.
    
Allerdings braucht man dazu eine Messeinrichtung (z.B. einen Shelly oder eine andere Messsteckdose) die den ins Haus gehenden Strom gegenmisst.

**Reply by Just-Zuul** [reply]
nochmal richtig formatiert:
```
service: pyscript.set_ef_powerstream_custom_load_power
data:
  SerialNumber: HWxxxxxxxxxxxxxxxxx
  TotalPower: >
    {% set tasmota_value = states('sensor.tasmota_lesekopf_power_curr') | int %}  
    {% set abzugswert = states('input_number.globaler_abzug') | int %}
    {% set schwelle = states('input_number.schwellwert') | int %}
    {% set inverter_ist = states('sensor.shelly_real_einspeisung_watt') | int %}
    {% set inverter_soll = states('sensor.inverter_output_watts') | int %}
    {% set ausgleichswert = states('sensor.inverter_other_loads') | int %}
    {% set inverter_diff = inverter_soll - inverter_ist %}
    {% set acceptable_range =
      7 if inverter_soll < 50 else
      10 if inverter_soll < 150 else
      15 if inverter_soll < 300 else
      20 if inverter_soll < 400 else
      25 %}
    {% if tasmota_value < schwelle %}
      {% set diff = schwelle - tasmota_value %}
      {% set neuer_abzugswert = abzugswert + diff %}
      {% if neuer_abzugswert > 800 %}
        {% set neuer_abzugswert = 800 %}
      {% elif neuer_abzugswert < 0 %}
        {% set neuer_abzugswert = 0 %}
      {% endif %}
      {{ tasmota_value - neuer_abzugswert }}
    {% else %}
      {{ tasmota_value - abzugswert }}
    {% endif %}
  Automation: true
enabled: true
```

**Reply by Neewieedo** [reply]
@Just-Zuul 

Ich denke Du hast mich falsch verstanden ;-) 
Ich wollte nur darauf, dass man die Berechnung nur an einer Stelle (HA oder im pyscript) machen sollte, sonst wird es glaub zu kompliziert...  
Da im Standard pyscript die Berechnung "fest" drin ist, muss man das dann halt ausblenden wenn man die Berechnung in HA
vornimmt.
Daher hab ich im _pyscript.set_ef_powerstream_custom_load_power_ die Berechnung auskommentiert:

```
    #CalPermanentWatts = cur_permanentWatts + TotalPower - TotalPowerOffSet
    CalPermanentWatts = TotalPower
```

Der Hinweis der Einspeisung nicht mit dem Cloud-Wert sondern mit meiner Tasmaota-Strommessdose zu machen war
richtig gut und wurde gleich umgesetzt.

**Frage:** Du hast nicht zufällig eine EcoFlow Batterie in das EnergieDashboard eingebunden und kannst mir einen Tipp dazu geben?

Ach und zu meinem Punkt 4 oben... Der Code zur Überwachung der Tibberwerte funktioniert jetzt:

```
alias: Tibber Status ermitteln
description: ""
trigger:
  - platform: state
    entity_id:
      - sensor.power_tibber
    for:
      hours: 0
      minutes: 2
      seconds: 0
    id: "1"
  - platform: state
    entity_id:
      - sensor.power_tibber
    to: unavailable
    id: "1"
  - platform: state
    entity_id:
      - sensor.power_tibber
condition: []
action:
  - if:
      - condition: trigger
        id:
          - "1"
    then:
      - metadata: {}
        data: {}
        target:
          entity_id: input_boolean.tibber_status
        action: input_boolean.turn_off
    else:
      - metadata: {}
        target:
          entity_id: input_boolean.tibber_status
        action: input_boolean.turn_on
        data: {}
mode: single
```

**Reply by Just-Zuul** [reply]
Dito - Du hast mich ebenso falsch verstanden ;-)

Du musst im pyscript gar nichts auskommentieren oder verändern - in dem Script ist, mit Ausnahme der Dev-Zugangsdaten, KEINE Änderung nötig!

Auch wenn Du die Cloudabfrage umgehst, indem Du den Wert des other_load-Sensors im Vorfeld abziehst musst Du gar nichts ändern.


Mit Tibber habe ich nix am Hut. Und Ja - ich habe meine Batterien mit integriert, ebenso die Steuerungen hierfür.
Daher, zu:

> "Frage: Du hast nicht zufällig eine EcoFlow Batterie in das EnergieDashboard eingebunden und kannst mir einen Tipp dazu geben?"

musst Du schon ein bisschen genauer werden.

Weiter oben siehst Du ja den Screenshot meines Dashboards

**Reply by Neewieedo** [reply]
Na, dann ist doch alles TipTop, frei nach dem Motto: Minus und Minus gibt Plus ;-)

**Comment by HansMeulenbroek** [comment]
Moin, bin auch hiermit beschäftigt. Hatte es vorige Woche mehr oder weniger laufend mit der US Version von der EcoFlow cloud (der war im HACS vorhanden). Habe neulich alles neu eingerichtet mit tolwi EcoFlow cloud aber bekomme es nicht am laufen. Wie kann ich Log-informationen von diesem Script finden/lesen?
![Scherm­afbeelding 2025-06-03 om 07 49 13](https://github.com/user-attachments/assets/619a371d-ee47-4944-838e-78d601dc1b57)
![Scherm­afbeelding 2025-06-03 om 07 47 55](https://github.com/user-attachments/assets/7787ddcf-d906-4a4e-8013-b478dfd8a042)

**Comment by HansMeulenbroek** [comment]
deleted a copy of my HA dashboard, had some info I forgot to hide

**Reply by Just-Zuul** [reply]
zum Einen: gib Deiner Automatisation einen ordentlichen Namen (alias), dann, bearbeite Deine Automatisation nicht in der YAML sondern über die Automatisationen - auch dort kannst Du im YAML Mode arbeiten aber du zerballerst Dir nicht geich alles wenn Du einen Fehler machst

Dann:  Schreibe mal einen Zahlenwert hinter TotalPower:   (zb.  TotalPower: 50)  Wenn er dann 50 ausgibt (sich auf 50 stellt, dann passt die Anbindung.. wenn nicht = Fehlersuche)

Du hast einen lk13be  - müsste es dann nicht power_curr sein ?? 
Check mal, ob Du den richtigen Sensor hast.

**Comment by HansMeulenbroek** [comment]
Thanks @Just-Zuul have it now working for some weeks [https://hansmeulenbroek.de/balkonkraftwerk.html](url) see my website for the details
